blockchain/chain/src/store/index.rs

```diff
@@ -0,0 +1,162 @@
+        if from.epoch() - to <= SKIP_LENGTH {
+            return self.walk_back(from, to).await;
+        }
+
+        let rounded = self.round_down(from).await?;
+
+        let mut cur = rounded.key().clone();
+        loop {
+            let entry = self.skip_cache.write().await.get(&cur).cloned();
+            let lbe = if let Some(cached) = entry {
+                cached
+            } else {
+                self.fill_cache(std::mem::take(&mut cur)).await?
+            };
+
+            if lbe.tipset.epoch() == to || lbe.parent_height < to {
+                return Ok(lbe.tipset.clone());
+            } else if to > lbe.target_height {
+                return self.walk_back(lbe.tipset.clone(), to).await;
+            }
+
+            cur = lbe.target.clone();
+        }
+    }
+
+    pub(crate) async fn get_tipset_by_height_without_cache(
+        &self,
+        from: Arc<Tipset>,
+        to: ChainEpoch,
+    ) -> Result<Arc<Tipset>, Error> {
+        self.walk_back(from, to).await
+    }
+
+    /// Fills cache with lookback entry, and returns inserted entry.
+    async fn fill_cache(&self, tsk: TipsetKeys) -> Result<Arc<LookbackEntry>, Error> {
+        let tipset = self.load_tipset(&tsk).await?;
+
+        if tipset.epoch() == 0 {
+            return Ok(Arc::new(LookbackEntry {
+                tipset,
+                parent_height: 0,
+                target_height: Default::default(),
+                target: Default::default(),
+            }));
+        }
+
+        let parent = self.load_tipset(tipset.parents()).await?;
+        let r_height = self.round_height(tipset.epoch()) - SKIP_LENGTH;
+
+        let parent_epoch = parent.epoch();
+        let skip_target = if parent.epoch() < r_height {
+            parent
+        } else {
+            self.walk_back(parent, r_height).await?
+        };
+
+        let lbe = Arc::new(LookbackEntry {
+            tipset,
+            parent_height: parent_epoch,
+            target_height: skip_target.epoch(),
+            target: skip_target.key().clone(),
+        });
+
+        self.skip_cache.write().await.put(tsk.clone(), lbe.clone());
+        Ok(lbe)
+    }
+
+    fn round_height(&self, height: ChainEpoch) -> ChainEpoch {
+        (height / SKIP_LENGTH) * SKIP_LENGTH
+    }
+    async fn round_down(&self, ts: Arc<Tipset>) -> Result<Arc<Tipset>, Error> {
+        let target = self.round_height(ts.epoch());
+
+        self.walk_back(ts, target).await
+    }
+
+    async fn walk_back(&self, from: Arc<Tipset>, to: ChainEpoch) -> Result<Arc<Tipset>, Error> {
+        if to > from.epoch() {
+            return Err(Error::Other(
+                "Looking for tipset with height greater than start point".to_string(),
+            ));
+        }
+
+        if to == from.epoch() {
+            return Ok(from);
+        }
+
+        let mut ts = from;
+        loop {
+            let pts = self.load_tipset(ts.parents()).await?;
+
+            if to > pts.epoch() {
```

Comment:
I'm guessing this to detect for empty epochs? Had to stare at that for a hot sec to see why this would ever happen lol

Reply:
yeah exactly haha